adding function for physical validation ensemble check

cwalker7 opened the pull request:
Here's my first pass at adding the physical validation ensemble check for replica exchange simulations. It extracts replica energies and converts to state energies in the same way as process_replica_exchange. I was thinking of incorporating this into process_replica_exchange itself, but decided to leave it separate for now since we will be changing how data is stored and passed to the analysis functions.

One temperature state is specified by the user or default, and the other is chosen using the formula with energy standard deviation. 

I tested this on the replica exchange example we've been using (12 replica, 50K to 600K), and am getting a deviation of ~3 quantiles from the analytical ratio. I will try other examples with longer trajectories.

cwalker7 committed:
adding function for physical validation ensemble check

mrshirts commented:
Nice!  Some comments:

* Why not just run validation on all of the neighboring pairs?  Should be cheap. Or potentially all the pairs, though that could start getting expensive, and ones that are too far away are not worth running. 

* 3 standard deviation is potentially something to worry about.  I got similar results for my simulation. Let's keep collecting data . . . Actually, when I set the initial point at a higher, then I get even more standard deviations (like up to 4 or 5).  Definitely something to look into.  Probably won't affect qualitative work, but will affect quantitative work. 

> I was thinking of incorporating this into process_replica_exchange itself

Maybe as an option to process_replica_exchange?

* I'm also getting a couple of errors: 

UserWarning: NVT with undefined natoms.
  warnings.warn(ensemble + ' with undefined natoms.')
/Users/mrshirts/anaconda3/lib/python3.6/site-packages/physical_validation/data/ensemble_data.py:88: 
UserWarning: NVT with undefined volume.

Volume is interesting, because the volume doesn't actually matter here; for NVT, you don't actually need the volume to run the check IIRC. Maybe you should check with Pascal on that.

* It's printing out a list at the very end - should it be doing that?

* "Overlap is 63.0% of trajectory"  I think it should be phrased something more like "trajectory 2 has 63% overlap with trajectory 1".  

"would be optimal". Maybe read "would be optimal for best statistics".  Any T would work, some give lower noise than others.

mrshirts commented on cg_openmm/simulation/physical_validation.py:
```diff
@@ -0,0 +1,161 @@
+    # Check that these are correct units and conversions to Gromacs units reference set
+    sim_data1.units = UnitData(
+        kb=kB._value,
+        energy_conversion=1,
```
You may not need to set all of these if 1 is the default for some of the values.

mrshirts commented on cg_openmm/simulation/physical_validation.py:
```diff
@@ -0,0 +1,161 @@
+from openmmtools.multistate import MultiStateReporter, MultiStateSampler, ReplicaExchangeSampler
+from openmmtools.multistate import ReplicaExchangeAnalyzer
+
+kB = (unit.MOLAR_GAS_CONSTANT_R).in_units_of(unit.kilojoule / (unit.kelvin * unit.mole))
```
if temperature list has units, should this be in units as well?

mrshirts replied:
Ah, I see that temps has units stripped off later.  Maybe kB should keep units . . . we'll have to think how to best do it. 

I did it a little different in expectations_fraction_contacts to try to keep track of the units a bit better.

mrshirts commented on cg_openmm/simulation/physical_validation.py:
```diff
@@ -0,0 +1,161 @@
+    # Find closest match
+    T_array = np.zeros(len(temperature_list))
+    for i in range(len(temperature_list)):
+        T_array[i] = temperature_list[i]._value
```
Probably should be using value_in_unit in most cases?

cwalker7 replied:
Ok I will update those - so value_in_unit just does an extra check to make sure it's in the units we want? We could also use that to avoid needing any conversions within physical validation (use value_in_unit with the Gromacs unit set)

mrshirts replied:
It actually gives the value in whatever units you specify as the argument.   _value gives the value in the units it is currently cast in, which is dangerous, since we generally don't know for sure what is passed in.

mrshirts commented on cg_openmm/simulation/physical_validation.py:
```diff
@@ -0,0 +1,161 @@
+        
+    # Check that these are correct units and conversions to Gromacs units reference set
+    sim_data2.units = UnitData(
+        kb=kB._value,
```
Probably only need to define the unit data once, since both sim_data1.units and sim_data2.units are assigned the same information.

cwalker7 commented:
Why not just run validation on all of the neighboring pairs?
- Yeah I think there's enough variation from pair to pair that we should check them all. We could make it an option to run all neighboring pairs, all pairs, or just a single pair (current algorithm).

About the errors:
- yeah these don't affect the results at all - I can bring this up on the physical validation issue page.

It's printing out a list at the very end - should it be doing that?
- We don't need that - just wanted to see what the returned 'quantiles' looked like

mrshirts commented:
> Yeah I think there's enough variation from pair to pair that we should check them all. We could make it an option to run all neighboring pairs, all pairs, or just a single pair (current algorithm).

Good plan.

> yeah these don't affect the results at all - I can bring this up on the physical validation issue page.

Yes, would be good to ask if that behavior is as intended.

> We don't need that - just wanted to see what the returned 'quantiles' looked like

OK, sounds good.

cwalker7 committed:
Merge branch 'master' of https://github.com/shirtsgroup/cg_openmm into add_physical_validation

Syncing with numerous updates.

cwalker7 committed:
added physical validation to test environment

cwalker7 committed:
updating unit handling

cwalker7 committed:
Updated to determine temperature list from nc file internally

cwalker7 committed:
removed unneccesary import

cwalker7 commented:
A bunch of updates here:

- Added options for single temperature pair with optimal spacing (original algorithm), all adjacent pairs, or all pairs (last one has been crashing due to poor overlap error)
- Temperature list now determined internally (as in process replica exchange) since we no longer store it in pickle files
- Updated units to use value_in_unit instead of _value, and only need to set UnitData() once
- Added physical validation (latest github version) to the testing environment file

cwalker7 committed:
don't need these imports either

mrshirts commented on cg_openmm/simulation/physical_validation.py:
```diff
@@ -0,0 +1,282 @@
+            temperature=T_array[state1_index]
+            )
+            
+        sim_data1.units = UnitData(
```
Can you just define a default UnitData object, then just state that sim_data1.units = default_UnitData?  Seems like that would reduce duplication / chance for errors - since they are all look like they are the same.

mrshirts commented on cg_openmm/simulation/physical_validation.py:
```diff
@@ -0,0 +1,282 @@
+                # State 1
+                state1_index = i
+
+                sim_data1 = SimulationData()
```
Could you write a single function that is used both in the single one, and loop over here, to reduce duplication?

cwalker7 replied:
Added an internal function to set the SimulationData given the state_energies, temperature array, and state indices. How's this look? Seems to be producing the same results as before for 'single' and 'adjacent' options, 'all' still failing due to the bin overlap error - I can make a separate issue for that.

mrshirts replied:
> Added an internal function to set the SimulationData given the state_energies, temperature array, and state indices.

That change makes a lot more sense.

> Seems to be producing the same results as before for 'single' and 'adjacent' options, 'all' still failing due to the bin overlap error - I can make a separate issue for that.

That sounds fine.  I would see if you can catch the error using try/except syntax for each state.

mrshirts reviewed (approved):
This looks fine to merge, but you might consider how to remove some code duplication as indicated in the line-by-line comments

cwalker7 committed:
adding set_simulation_data function to reduce repititive code
